### PR TITLE
Refactor library management

### PR DESCRIPTION
**Pull request by diml**

### CODE REVIEW

**Comment by diml** [comment]
This took much longer than I expected but it's now working. I think things are a bit cleaner now, in particular all the library resolution and management is done at a single place.

Here is a summary of the changes:

## Lib module

We have a new module `Lib` that replaces `Lib`, parts of `Lib_db` and parts of `Findlib`. It is used to manage all libraries (internal and extrernal).  `Lib.t` represent a completely resolved library, i.e. where all the dependencies have been resolved. `Lib.Compile` is used to provide what is necessary to build the library itself. `Lib.Meta` provides what is necessary to generate the `META` file for the library.

We also have library databases represented as `Lib.DB.t`. A library database is simply a mapping from names to `Lib.t` values and and created from a `resolve` function that looks up a name and return a `Lib.Info.t`. A `Lib.Info.t` is the same as a `Lib.t` except that dependencies are not resolved.

A library database can have a parent database that is used to lookup names that are not found in the current database. In practice we have the following hierarchy:

1. For every scope, we have a library database that holds all the libraries of this scope. In this DB, a library can be referred by either it's name or public name
2. the parent of each of these databases is a database that holds all the public libraries of the workspace. In this DB libraries must be referred by their public name
3. the parent of this DB is for installed libraries

- (1) databases are accessible via `Scope.libs (Super_context.find_scope_by_{name,dir} sctx xxx)`
- (2) is accessible via `Super_context.public_libs sctx`
- (3) is accessible via `Super_context.installed_libs sctx`

The dependencies of a library are always resolved inside the DB it is part of. When we compute a transitive closure, we check that we don't have two libraries from two different  DB with the same name. So for instance linting Base should now supported.

## Jbuild.Scope_info

`Jbuild.Scope` was renamed `Jbuild.Scope_info`

## Scope module

This replaces `Lib_db`. A `Scope.t` is now just a pair of a `Jbuild.Scope_info.t` and a `Lib.DB.t`.  `Scope.DB.t` is an object used to lookup scopes by either name or directory.

We no longer have an external scope or special anonymous scope. Instead one should use `Super_context.installed_libs` or `Super_context.public_libs` depending on the context.

File: src/build_system.ml [hunk 1]
```diff
@@ -463,10 +463,14 @@ module Build_exec = struct
...
-          dyn_deps := Pset.union !dyn_deps !dyn_deps';
-          x
+          match exec dyn_deps' m.t x with
+          | x ->
+            m.state <- Evaluated (x, !dyn_deps');
+            dyn_deps := Pset.union !dyn_deps !dyn_deps';
+            x
+          | exception exn ->
+            m.state <- Unevaluated;
+            reraise exn
```
Comment (rgrinberg) [inline src/build_system.ml]: Is this some bug that's being fixed? It doesn't seem related to the lib refactoring.

File: src/odoc.ml [hunk 2]
```diff
@@ -157,19 +157,37 @@ let css_file ~doc_dir = doc_dir ++ "odoc.css"
...
       ~requires ~(dep_graphs:Ocamldep.Dep_graph.t Ml_kind.Dict.t) =
-  let doc_dir = SC.Doc.dir sctx (dir, lib) in
-  let obj_dir = Lib.lib_obj_dir dir lib in
-  let lib_unique_name = SC.unique_library_name sctx (Lib.internal (dir, lib)) in
+  let doc_dir = SC.Doc.dir sctx lib in
+  let obj_dir, lib_unique_name =
+    match Lib.DB.find (Scope.libs scope) lib.name with
+    | Error Not_found -> assert false
+    | Error (Hidden _) ->
+      (Utils.library_object_directory ~dir lib.name,
```
Comment (rgrinberg) [inline src/odoc.ml]: Would you mind explaining this case a bit? I'm not sure when an internal lib would be hidden. Also, it's still unsatisfying to go outside of the lib interface to get the object dir in that case.

File: src/odoc.ml [hunk 3]
```diff
@@ -267,17 +285,14 @@ let gen_rules sctx ~dir:_ rest =
...
-    let scope =
-      { With_required_by.
-        data = scope
-      ; required_by = [Alias (Path.of_string "doc")]
-      } in
-    let open Option.Infix in
-    Option.iter (Lib_db.Scope.find scope lib >>= Lib.src_dir)
-      ~f:(fun dir -> SC.load_dir sctx ~dir)
+    match Lib.DB.find lib_db lib with
+    | Error _ -> ()
```
Comment (rgrinberg) [inline src/odoc.ml]: A bit off topic since that's the old behavior anyway, but why are we ignoring such errors? This cames seems impossible, so shouldn't it be assert false or a Sexp.code_error?

File: src/result/result.mli [hunk 4]
```diff
@@ -13,8 +13,18 @@ include sig
...
+
+module O : sig
+  val ( >>| ) : ('a, 'error) t -> ('a -> 'b) -> ('b, 'error) t
+  val ( >>= ) : ('a, 'error) t -> ('a -> ('b, 'error) t) -> ('b, 'error) t
+end
+
+val map  : ('a, 'error) t -> f:('a -> 'b) -> ('b, 'error) t
+val bind : ('a, 'error) t -> f:('a -> ('b, 'error) t) -> ('b, 'error) t
+
+val map_error : ('a, 'error1) t -> f:('error1 -> 'error2) -> ('a, 'error2) t
```
Comment (rgrinberg) [inline src/result/result.mli]: Rather than extending this module, we should consider just vendoring rresult.ml I think.

File: bin/main.ml [hunk 5]
```diff
@@ -758,7 +756,7 @@ let external_lib_deps =
                in
                let missing =
                  String_map.filter externals ~f:(fun name _ ->
-                   not (Findlib.available context.findlib name))
+                   not (Findlib.mem context.findlib name))
```
[the remark this answers is not in the document]
Comment (diml) [reply bin/main.ml]: Ok, makes sense. This is the same for Lib.DB.mem actually

File: src/build_system.ml (hunk 1 above)
Comment (diml) [reply src/build_system.ml]: Yh, it just that it popped up in a test with this PR. Basically if evaluating a memoized Build.t failed, it was staying forever in `Evaluating` state, causing a cycle error

File: src/odoc.ml (hunk 2 above)
Comment (diml) [reply src/odoc.ml]: It's for libraries that are optional with unavailable dependencies.

> Also, it's still unsatisfying to go outside of the lib interface to get the object dir in that case.

Indeed. I suppose we could keep the `Info.t` in this case

File: src/odoc.ml (hunk 3 above)
Comment (diml) [reply src/odoc.ml]: Actually it's not impossible. We can reach this case with:

```sh
$ jbuilder build _build/default/_doc/blah/...
```

Currently, this will cause a `no rule found for ...` error. Maybe we could return something more precise

File: bin/main.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (diml) [reply bin/main.ml]: done

File: src/odoc.ml (hunk 2 above)
Comment (diml) [reply src/odoc.ml]: done

File: src/result/result.mli (hunk 4 above)
Comment (diml) [reply src/result/result.mli]: I think we should limit the vendoring to things to things that provide complex features or that are dependencies of things we vendor. `Result` is a very simple module. Moreover the API conventions of rresult are quite different from the rest of the Dune codebase
